### PR TITLE
Run `integrity_test` in Externalities

### LINKED ISSUE #13726: Run `integrity_test` in externalities
We should probably run *all* of then inside externalities. In the past I was concerned about unintended consequences, but it just annoyed us again in: https://github.com/paritytech/polkadot/pull/6957

Since there a parameter type was configured to be `storage` instead of `const`, so it would fail outside externalities.

### PR DESCRIPTION
**Pull request by ggwpez**
Closes https://github.com/paritytech/substrate/issues/13726

Changes:
- Always run `integrity_test` in test externalities.

### CODE REVIEW

File: frame/support/src/lib.rs [hunk 1]
```diff
@@ -65,6 +65,13 @@ pub use sp_std;
 #[doc(hidden)]
 pub use tt_call::*;
 
+// Re-exports go here:
+#[doc(hidden)]
+pub mod private {
```
Comment (bkchr) [inline frame/support/src/lib.rs:70]:
```suggestion
pub mod __private {
```

Maybe to make it even more "weird" to use this :P

File: frame/support/procedural/src/pallet/expand/hooks.rs [hunk 2]
```diff
@@ -251,11 +251,16 @@ pub fn expand_hooks(def: &mut Def) -> proc_macro2::TokenStream {
 			for #pallet_ident<#type_use_gen> #where_clause
 			{
 				fn integrity_test() {
-					<
-						Self as #frame_support::traits::Hooks<
-						<T as #frame_system::Config>::BlockNumber
-						>
+					let ext = #frame_support::sp_io::TestExternalities::default();
+					let mut ro_ext = #frame_support::private::ReadOnlyExternalities::from(&ext.backend);
```
Comment (bkchr) [inline frame/support/procedural/src/pallet/expand/hooks.rs:255]: Why read only?

File: frame/system/benchmarking/src/lib.rs [hunk 3]
```diff
@@ -18,6 +18,7 @@
 // Benchmarks for Utility Pallet
 
 #![cfg_attr(not(feature = "std"), no_std)]
+#![cfg(feature = "runtime-benchmarks")]
```
Comment (bkchr) [inline frame/system/benchmarking/src/lib.rs:21]: Weird that the pallet even has this feature as this pallet should only be used when `runtime-benchmarks` are enabled somewhere else.

File: frame/support/procedural/src/pallet/expand/hooks.rs (hunk 2 above)
Comment (ggwpez) [reply frame/support/procedural/src/pallet/expand/hooks.rs:255]: My thinking was that this should make developers aware that storage changes are being discarded. Or do you think its not needed?

File: frame/system/benchmarking/src/lib.rs (hunk 3 above)
Comment (ggwpez) [reply frame/system/benchmarking/src/lib.rs:21]: I was just running the tests for all pallets and one of the benchmarking tests failed. So it was "enabled" manually by `cargo test`.

File: frame/support/procedural/src/pallet/expand/hooks.rs (hunk 2 above)
Comment (bkchr) [reply frame/support/procedural/src/pallet/expand/hooks.rs]: I mean  this is a test as the name says. Maybe we should improve the docs. However, there could be someone calling a "real" function for the test that also writes something.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3166352

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3167123

**Review by bkchr — APPROVED** [review]

**Comment by ggwpez** [comment]
bot rebase

File: frame/support/src/dispatch.rs [hunk 4]
```diff
@@ -895,8 +895,8 @@ impl<T> PaysFee<T> for (u64, Pays) {
 /// * `offchain_worker`: Executes at the beginning of a block and produces extrinsics for a future
 ///   block upon completion. Using this function will implement the
 ///   [`OffchainWorker`](./traits/trait.OffchainWorker.html) trait.
-/// * `integrity_test`: Executes in a test generated by `construct_runtime`, note it doesn't execute
-///   in an externalities-provided environment. Implement
+/// * `integrity_test`: Executes in a test generated by `construct_runtime`, note it does execute it
+/// * in an externalities-provided environment. Implement
```
Comment (muharem) [inline frame/support/src/dispatch.rs:899]:
```suggestion
///   in an externalities-provided environment. Implement
```

**Review by muharem — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3206451

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3206452

**Comment by ggwpez** [comment]
bot merge

**Comment by ggwpez** [comment]
bot rebase

File: frame/system/benchmarking/src/lib.rs (hunk 3 above)
Comment (xlc) [reply frame/system/benchmarking/src/lib.rs:21]: without this, my build isn't working. can we have this ported to the 1.0.0 release branch @bkchr

Comment (xlc) [reply frame/system/benchmarking/src/lib.rs:21]: this is the correct solution https://github.com/paritytech/substrate/pull/14617

Comment (xlc) [reply frame/system/benchmarking/src/lib.rs:21]: We need this for Acala to compile with polkadot-v1.0.0 branch

**Pull request merged**